### PR TITLE
Merge tuple into a single type representing same values.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 1]
```diff
@@ -76,11 +76,11 @@ private sealed class InProgressState : CompilationTrackerState
                 /// The list of changes that have happened since we last computed a compilation. The oldState corresponds to
                 /// the state of the project prior to the mutation.
                 /// </summary>
-                public ImmutableList<(ProjectState oldState, CompilationAndGeneratorDriverTranslationAction action)> PendingTranslationSteps { get; }
+                public ImmutableList<TranslationAction> PendingTranslationActions { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:79]: 1. i always found it unpleasant to hve to carry around this tuple and keep in sync.  So i pushed `ProjectState oldState` into `CompilationAndGeneratorDriverTranslationAction`
2. Rename CompilationAndGeneratorDriverTranslationAction to `TranslationAction`.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 2]
```diff
@@ -162,28 +161,28 @@ state switch
...
                     {
-                        var (priorState, priorAction) = pendingTranslationSteps.Last();
+                        var priorAction = pendingTranslationActions.Last();
                         var mergedTranslation = translate.TryMergeWithPrior(priorAction);
                         if (mergedTranslation != null)
                         {
                             // We can replace the prior action with this new one
-                            return pendingTranslationSteps.SetItem(
-                                pendingTranslationSteps.Count - 1,
-                                (oldState: priorState, mergedTranslation));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:181]: note the logic here.  when merging two actions, the `oldState` we pass in was the `oldState` paired with the translation step in the list.  thsi will be relevant later.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 3]
```diff
@@ -37,19 +41,25 @@ public override async Task<Compilation> TransformCompilationAsync(Compilation ol
...
-                public override CompilationAndGeneratorDriverTranslationAction? TryMergeWithPrior(CompilationAndGeneratorDriverTranslationAction priorAction)
+                public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
                 {
                     if (priorAction is TouchDocumentAction priorTouchAction &&
                         priorTouchAction._newState == _oldState)
                     {
-                        return new TouchDocumentAction(priorTouchAction._oldState, _newState);
+                        // As we're merging ourselves with the prior touch action, we want to keep the old project state
+                        // that we are translating from.
+                        return new TouchDocumentAction(priorAction.OldProjectState, priorTouchAction._oldState, _newState);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:51]: this preserves the sematnics that after merging, our 'oldstate' was the original old state of the prior action.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 4]
```diff
@@ -62,12 +72,14 @@ internal sealed class TouchAdditionalDocumentAction(AdditionalDocumentState oldS
...
-                public override CompilationAndGeneratorDriverTranslationAction? TryMergeWithPrior(CompilationAndGeneratorDriverTranslationAction priorAction)
+                public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
                 {
                     if (priorAction is TouchAdditionalDocumentAction priorTouchAction &&
                         priorTouchAction._newState == _oldState)
                     {
-                        return new TouchAdditionalDocumentAction(priorTouchAction._oldState, _newState);
+                        // As we're merging ourselves with the prior touch action, we want to keep the old project state
+                        // that we are translating from.
+                        return new TouchAdditionalDocumentAction(priorAction.OldProjectState, priorTouchAction._oldState, _newState);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:82]: this preserves the sematnics that after merging, our 'oldstate' was the original old state of the prior action.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 5]
```diff
@@ -279,10 +323,14 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver _)
...
+                public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
                 {
-                    // If the prior action is also a ReplaceGeneratorDriverAction, we'd entirely overwrite it's changes, so we can drop the prior one entirely.
-                    return priorAction is ReplaceGeneratorDriverAction ? this : null;
+                    // If the prior action is also a ReplaceGeneratorDriverAction, we'd entirely overwrite it's changes,
+                    // so we can drop the prior one's generator driver entirely.  Note: we still want to use it's
+                    // `OldProjectState` as that still represents the prior state we're translating from.
+                    return priorAction is ReplaceGeneratorDriverAction
+                        ? new ReplaceGeneratorDriverAction(oldGeneratorDriver, priorAction.OldProjectState, newProjectState)
+                        : null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:333]: this preserves the sematnics that after merging, our 'oldstate' was the original old state of the prior action.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction.cs [hunk 6]
```diff
@@ -10,11 +10,27 @@ namespace Microsoft.CodeAnalysis;
...
     {
+        /// <summary>
+        /// The original state of the project prior to the user edit.
+        /// </summary>
+        public readonly ProjectState OldProjectState;
+
+        /// <summary>
+        /// The state of the project after the user edit was made.
+        /// </summary>
+        public readonly ProjectState NewProjectState;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction.cs:26]: adding both of these to the translation action simplified a lot of code.  we have code that would pass along the old-state and action, and code that compute the new-state and action.  Now, one abstraction works for all those cases.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 7]
```diff
@@ -257,7 +305,11 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
                 }
             }
 
-            internal sealed class ReplaceGeneratorDriverAction(GeneratorDriver oldGeneratorDriver, ProjectState newProjectState) : CompilationAndGeneratorDriverTranslationAction
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:260]: a few actions were already passing along the newProjectState.  so it's not a stretch to just have all of htem do this.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 8]
```diff
@@ -1213,63 +1219,55 @@ public SolutionCompilationState AddDocuments(ImmutableArray<DocumentInfo> docume
     {
         return AddDocumentsToMultipleProjects(documentInfos,
             static (documentInfo, project) => project.CreateDocument(documentInfo, project.ParseOptions, new LoadTextOptions(project.ChecksumAlgorithm)),
-            static (oldProject, documents) => (oldProject.AddDocuments(documents), new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction(documents)));
+            static (oldProject, documents) => new TranslationAction.AddDocumentsAction(oldProject, oldProject.AddDocuments(documents), documents));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1222]: here was code that had to return tuples of 'new project state' and the action.  now, we just pass the new project state into the action.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 9]
```diff
@@ -248,29 +247,21 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
...
-                }
-                else
-                {
+                    ? new TranslationAction.TouchDocumentAction(oldProjectState, oldProjectState.UpdateDocument(docState, contentChanged: true), oldState, docState)
                     // The document wasn't present in the original snapshot at all, and we just need to add the
                     // document.
-                    pendingActions = [(oldState: inProgressProject, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
-                    inProgressProject = inProgressProject.AddDocuments([docState]);
-                }
+                    : new TranslationAction.AddDocumentsAction(oldProjectState, oldProjectState.AddDocuments([docState]), [docState]);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:264]: i really like how much simpler this code gets.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @ToddGrun this is ready for review.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 10]
```diff
@@ -144,10 +143,10 @@ SymbolKind.NetModule or
                         _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
                     };
 
-                    var finalSteps = UpdatePendingTranslationSteps(
+                    var finalSteps = UpdatePendingTranslationActions(
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:146]: >finalSteps [](http://example.com/codeflow?start=24&length=10)

finalActions?

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 11]
```diff
@@ -184,7 +214,11 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
                 }
             }
 
-            internal sealed class ProjectAssemblyNameAction(string assemblyName) : CompilationAndGeneratorDriverTranslationAction
+            internal sealed class ProjectAssemblyNameAction(
+                ProjectState oldProjectState,
+                ProjectState newProjectState,
+                string assemblyName)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:220]: Something to think about: in this case you don't really need to capture this as a separate property -- it's now just stored by the newProjectState and you could just grab the property from it too. The document actions you still need to remember which document(s) it is, but here you don't.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 12]
```diff
@@ -162,28 +161,28 @@ state switch
...
                     // We have a translation action; are we able to merge it with the prior one?
-                    if (!pendingTranslationSteps.IsEmpty)
+                    if (!pendingTranslationActions.IsEmpty)
                     {
-                        var (priorState, priorAction) = pendingTranslationSteps.Last();
+                        var priorAction = pendingTranslationActions.Last();
                         var mergedTranslation = translate.TryMergeWithPrior(priorAction);
                         if (mergedTranslation != null)
                         {
                             // We can replace the prior action with this new one
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:177]: Consider adding a few asserts here that the oldState and newState of the merged action matches the appropriate old/new states from the original two actions.

**Review by jasonmalinowski — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:220]: good point!

**Pull request merged**